pkg/rbac/rbac.go

```diff
@@ -0,0 +1,314 @@
+// set on the AccessReviewer, to create a new k8s client. If k8sConfig is not available,
+// then the configured k8s client is returned.
+func (r *AccessReviewer) getKubeClientForUser(userToken string) (kubernetes.Interface, error) {
+	if r.kubeConfig != nil {
+		// if a valid userToken
+		if userToken != "" {
+			// make a copy of the RestConfig to avoid overwrites when multiple api calls are made in parallel
+			userKubeConfig := rest.CopyConfig(r.kubeConfig)
+			userKubeConfig.BearerToken = userToken
+			// create the clientset
+			kclient, err := kubernetes.NewForConfig(userKubeConfig)
+			if err != nil {
+				return nil, err
+			}
+
+			return kclient, nil
+		}
+
+		return nil, fmt.Errorf(
+			"failed to get a client to connect to the kubernetes cluster:" +
+				"When KubeConfig is provided, a valid userToken must be set on all access review calls")
+	}
+
+	// if kubeConfig isnt set then return the kubeClient set
+	return r.kubeClient, nil
+}
+
+// GetMetricsAccess retrieves the user's ACLs from the k8s cluster  and processes them to determine
+// user's access to observability metrics that are gathered from managed clusters.
+// It returns a map of allowed managedclusters and the allowed namespaces for each managedcluster.
+//
+// - userToken is the user's OAuth bearer token, is required if k8s config was set on the AccessReviewer
+//
+// - clusters are the  names of the managedclusters for which  allowed metrics access is processed and returned.
+// if no clusters are specified, then  metrics access is returned for all "allowed" managedclusters.
+func (r *AccessReviewer) GetMetricsAccess(userToken string, clusters ...string) (map[string][]string, error) {
+	klog.V(2).Infof("GetMetricsAccess for clusters: %v", clusters)
+
+	// get Client to talk to the Kubernetes cluster
+	userKClient, err := r.getKubeClientForUser(userToken)
+	if err != nil {
+		return nil, err
+	}
+
+	// get all user ACLs on ManagedCluster resources
+	resourceACLs, err := GetResourceAccess(userKClient, MetricsACLConfig.groupRes, clusters, "")
+	if err != nil {
+		return nil, err
+	}
+
+	klog.V(2).Infof(" resource access results: %v", resourceACLs)
+
+	// from the list of all ACLs for ManagedCluster, filter out the "metrics" specific acls and grab the namespaces
+	metricsAccessResults := make(map[string][]string)
+
+	for clustername, clusteracls := range resourceACLs {
+		klog.V(2).Infof("clustername [%s] acls[%s]\n", clustername, clusteracls)
+
+		// list of namespaces for the cluster
+		metricsAccessMap := make(map[string]bool)
+
+		for _, acl := range clusteracls {
+			// filter verbs that start with metrics/ and grab the namespace
+			// if verb is set to *, set namespace to * to indicate access to all namespaces
+			if strings.HasPrefix(acl, MetricsACLConfig.verb) {
+				metricsAccessMap[strings.TrimPrefix(acl, MetricsACLConfig.verb)] = true
+			} else if acl == "*" {
+				metricsAccessMap["*"] = true
+			}
+		}
+
+		nsWithMetricsAccess := []string{}
```

Comment:
Optional: This is a good practice to avoid `append` unnecessarily recreating the underlying array as it grows.
```suggestion
		nsWithMetricsAccess := make([]string{}, 0, len(metricsAccessMap))
```